File: nim/src/minsk/codeAnalysis/syntaxFacts.nim
import syntaxKind

proc binaryOperatorPrecedence*(kind: SyntaxKind): int =
  case kind
  of SyntaxKind.StarToken, SyntaxKind.SlashToken:
    2
  of SyntaxKind.PlusToken, SyntaxKind.MinusToken:
    1
  else:
    0

File: hs/src/Minsk/SyntaxFacts.hs
module Minsk.SyntaxFacts
  ( Precedence (..),
    binaryOperatorPrecedence,
    unaryOperatorPrecedence,
  )
where

import BasicPrelude
import Minsk.SyntaxKind (SyntaxKind (..))

data Precedence = PZero | PTerm | PFactor | PUnary
  deriving (Eq, Ord)

binaryOperatorPrecedence :: SyntaxKind -> Precedence
binaryOperatorPrecedence k = case k of
  PlusToken -> PTerm
  MinusToken -> PTerm
  StarToken -> PFactor
  SlashToken -> PFactor
  _ -> PZero

unaryOperatorPrecedence :: SyntaxKind -> Precedence
unaryOperatorPrecedence k = case k of
  PlusToken -> PUnary
  MinusToken -> PUnary
  _ -> PZero
